From d8f38c1380c3cef1b0d12ef4ef17711bb4a3d985 Mon Sep 17 00:00:00 2001
From: vincentliu <vliu@mozilla.com>
Date: Wed, 19 Feb 2014 15:27:47 +0800
Subject: [PATCH] Modify
 0008-Bug-957509-Reduce-USS-of-Nuwa-and-preallocated-proce
 due to patch fail

---
 dom/ipc/ContentChild.cpp  |    9 ++++++++-
 dom/ipc/ContentParent.cpp |   17 +++++++----------
 dom/ipc/ContentParent.h   |    5 ++---
 3 files changed, 17 insertions(+), 14 deletions(-)

diff --git a/dom/ipc/ContentChild.cpp b/dom/ipc/ContentChild.cpp
index 42d4023..7a81859 100644
--- a/dom/ipc/ContentChild.cpp
+++ b/dom/ipc/ContentChild.cpp
@@ -721,6 +721,10 @@ ContentChild::RecvPBrowserConstructor(PBrowserChild* actor,
         MOZ_ASSERT(!sFirstIdleTask);
         sFirstIdleTask = NewRunnableFunction(FirstIdle);
         MessageLoop::current()->PostIdleTask(FROM_HERE, sFirstIdleTask);
+
+        // GC is disabled in the preallocated process. Enable it when we start
+        // loading as the browser or app.
+        nsJSContext::EnableGC();
     }
 
     return true;
@@ -1640,7 +1644,10 @@ public:
 
         // Perform other after-fork initializations.
         InitOnContentProcessCreated();
-
+        
+        // Don't perform GC in the preallocated process to avoid touching
+        // COW pages.
+        nsJSContext::DisableGC();
         return NS_OK;
     }
 };
diff --git a/dom/ipc/ContentParent.cpp b/dom/ipc/ContentParent.cpp
index 042fb3c..84ae759 100644
--- a/dom/ipc/ContentParent.cpp
+++ b/dom/ipc/ContentParent.cpp
@@ -1287,8 +1287,7 @@ ContentParent::ContentParent(mozIApplication* aApp,
     Open(mSubprocess->GetChannel(), mSubprocess->GetOwnedChildProcessHandle());
 
     InitInternal(aInitialPriority,
-                 true, /* Setup off-main thread compositing */
-                 true  /* Send registered chrome */);
+                 true /* aIsTemplate */);
 }
 
 #ifdef MOZ_NUWA_PROCESS
@@ -1360,8 +1359,7 @@ ContentParent::ContentParent(ContentParent* aTemplate,
     }
 
     InitInternal(priority,
-                 false, /* Setup Off-main thread compositing */
-                 false  /* Send registered chrome */);
+                 false /* aIsTemplate */);
 }
 #endif  // MOZ_NUWA_PROCESS
 
@@ -1393,8 +1391,7 @@ ContentParent::~ContentParent()
 
 void
 ContentParent::InitInternal(ProcessPriority aInitialPriority,
-                            bool aSetupOffMainThreadCompositing,
-                            bool aSendRegisteredChrome)
+    bool aIsTemplate)
 {
     // Set the subprocess's priority.  We do this early on because we're likely
     // /lowering/ the process's CPU and memory priority, which it has inherited
@@ -1404,7 +1401,7 @@ ContentParent::InitInternal(ProcessPriority aInitialPriority,
     // must come after the Open() call above.
     ProcessPriorityManager::SetProcessPriority(this, aInitialPriority);
 
-    if (aSetupOffMainThreadCompositing) {
+    if (aIsTemplate) {
         // NB: internally, this will send an IPC message to the child
         // process to get it to create the CompositorChild.  This
         // message goes through the regular IPC queue for this
@@ -1425,7 +1422,7 @@ ContentParent::InitInternal(ProcessPriority aInitialPriority,
         }
     }
 
-    if (aSendRegisteredChrome) {
+    if (aIsTemplate) {
         nsCOMPtr<nsIChromeRegistry> registrySvc = nsChromeRegistry::GetService();
         nsChromeRegistryChrome* chromeRegistry =
             static_cast<nsChromeRegistryChrome*>(registrySvc.get());
@@ -1434,7 +1431,7 @@ ContentParent::InitInternal(ProcessPriority aInitialPriority,
 
     mMessageManager = nsFrameMessageManager::NewProcessMessageManager(this);
 
-    if (gAppData) {
+    if (gAppData && aIsTemplate) {
         nsCString version(gAppData->version);
         nsCString buildID(gAppData->buildID);
         nsCString name(gAppData->name);
@@ -1445,7 +1442,7 @@ ContentParent::InitInternal(ProcessPriority aInitialPriority,
     }
 
     nsStyleSheetService *sheetService = nsStyleSheetService::GetInstance();
-    if (sheetService) {
+    if (sheetService && aIsTemplate) {
         // This looks like a lot of work, but in a normal browser session we just
         // send two loads.
 
diff --git a/dom/ipc/ContentParent.h b/dom/ipc/ContentParent.h
index 7d6a3f0..27ce74a 100644
--- a/dom/ipc/ContentParent.h
+++ b/dom/ipc/ContentParent.h
@@ -274,10 +274,9 @@ private:
     // The common initialization for the constructors.
     void InitializeMembers();
 
-    // The common initialization logic shared by all constuctors.
+    // The common initialization logic shared by both constuctors.
     void InitInternal(ProcessPriority aPriority,
-                      bool aSetupOffMainThreadCompositing,
-                      bool aSendRegisteredChrome);
+                      bool aIsTemplate);
 
     virtual ~ContentParent();
 
-- 
1.7.9.5

